Voting system

TurtIeSocks opened the pull request:
- Upvote/Downvote Buttons
- Green when upvoted
- Red when downvoted
- User can only upvote or downvote once
- Automatically adjusts the total points based off of what each user selects

TurtIeSocks committed:
Initial

Co-Authored-By: Tom Low <62975710+tomlow@users.noreply.github.com>

TurtIeSocks committed:
Finished Vote System

- Upvote/Downvote Buttons
- Green when upvoted
- Red when downvoted
- User can only upvote or downvote once
- Automatically adjusts the total points based off of what each user selects

Co-Authored-By: Tom Low <62975710+tomlow@users.noreply.github.com>

TurtIeSocks committed:
Merge branch 'main' into voting-system

TurtIeSocks committed:
Update podcastVotesRouter.js

tomlow committed:
resolved conflcts

tomlow committed:
Merge branch 'voting-system' of https://github.com/TurtIeSocks/launch31-group-project into voting-system

TurtIeSocks committed:
Formatting

nwalberts commented on client/src/components/podcasts/PodcastTile.js:
```diff
@@ -1,20 +1,167 @@
+    } catch (error) {
+      console.error(error.message)
+    }
+  }
```
you've done great work, but this will significantly slow down the load time on the page and lead to unecessary re-renders. Rather than fetching on each tile (if we had 20 tiles, there would be 20 new fetch requests) we should serialize votes with the podcasts that have been voted for.

It should be mentioned (but this is fine) that votes were also supposed to be on reviews, not the podcasts themselves :P

nwalberts replied:
I would save this refactor to last, and get to the others first

nwalberts commented on client/src/components/podcasts/PodcastTile.js:
```diff
@@ -1,20 +1,167 @@
+    } catch (error) {
+      console.error(error.message)
+    }
+  }
```
this is great, though it would have been allowed to combine newVote and editVote through one endpoint

nwalberts commented on server/src/db/Seeder.js:
```diff
@@ -5,11 +5,17 @@ import { User } from '../models/index.js'
+      email: 'test2@test.com',
+      cryptedPassword: 'hithere',
+      username: 'test2'
+    })
```
i would stick with the pattern you have for other seeders: use a UserSeeder

TurtIeSocks replied:
Resolved this in the branch we've been working on today that's based off of this branch so I'm going to leave it for now to reduce conflicts.

nwalberts commented on server/src/db/migrations/20210202161433_addVotesTable.cjs:
```diff
@@ -0,0 +1,32 @@
+      .index()
+      .unsigned()
+      .references('users.id')
+    table.integer('value')
```
hmmm I would probably have this has notNullable

nwalberts commented on server/src/models/Podcast.js:
```diff
@@ -45,6 +46,18 @@ class Podcast extends Model {
+            to: 'votes.userId'
+          },
+          to: 'users.id'
+        }
```
Unfortunately, this relation is no beuno for a few reasons, and should be removed. 

- you never use this functionality, there is no `podcast.$relatedQuery("votes")` (which is good, I like your votesRouter implementation actually) 
- the relation itself is wrong. The relationMapoing designates that you are trying to retrieve a modelClass of `Vote` and have dubbed the relation `votes`, but the join is representative of a many to many from podcasts to users _through_ votes
- even if you changed this to a many to many between users and podcasts, why would we want to know which users have voted for a podcast? if there was a feature that demanded this, then I could see the value

nwalberts replied:
additionally, I see no relation between the key players in your feature: podcasts and votes. You don't need to add it based on your current functionality, but that would have been far preferable

nwalberts commented on server/src/models/User.js:
```diff
@@ -72,6 +72,18 @@ class User extends uniqueFunc(Model) {
+            to: 'votes.podcastId'
+          },
+          to: 'podcasts.id'
+        }
```
same comment as what I said on podcasts. please remove this relation

nwalberts commented on server/src/models/Vote.js:
```diff
@@ -0,0 +1,32 @@
+          from: "votes.userId",
+          to: "users.id"
+        }
+      },
```
this one works, but the naming is improper. `users` should be `user`, since a vote can only have one user

nwalberts commented on server/src/models/Vote.js:
```diff
@@ -0,0 +1,32 @@
+        join: {
+          from: "votes.podcastId",
+          to: "podcasts.id"
+        }
```
I see your intention with this relation, but am concerned that you aren't thinking of it consciously
- a vote belongs to one podcast, you cant vote for two podcasts in one stroke. the relation should be a `BelongstoOne`
- `podcasts` should be `podcast` due to the BelongsToOne` relation

nwalberts commented on server/src/routes/api/v1/podcastVotesRouter.js:
```diff
@@ -0,0 +1,105 @@
+    votes.totalVotes = await Vote.query()
+      .where('podcastId', podcastId)
+      .sum('value as value')
+      .first()
```
nice job with sum. do we really need the `.first()` here? if so that's fine, but it does look odd. Ensure that this is returning one number that represents the total votes for a podcasts.

TurtIeSocks replied:
The `.first()` returns only an object instead of an "array" of one object. Not "essential" but made sense for what we're trying to accomplish.

nwalberts commented on server/src/routes/api/v1/podcastVotesRouter.js:
```diff
@@ -0,0 +1,105 @@
+    votes.userVotes = await Vote.query()
+      .where('podcastId', podcastId)
+      .andWhere('userId', userId)
+      .first()
```
these where statements are unnecessary. Any time you have to call `first()` at the end of a query, you likely want to be using `findOne` instead of `.where`

nwalberts commented on server/src/routes/api/v1/podcastVotesRouter.js:
```diff
@@ -0,0 +1,105 @@
+    newVote.userVotes = await Vote.query()
+      .insert({ podcastId, userId, value })
+      .first()
+      .returning('*')
```
wait...why are we calling first here? I believe you should delete line 42
https://vincit.github.io/objection.js/recipes/returning-tricks.html
you only need `first` if you have a where statement generally.

Ill allow line 43, since its best practice with PG databases, though this will break with any other database. generally though, we want to stick to one pattern. So if you have `insertAndFetch` elsewhere in your codebase (which is the alternative to using `returning`) you would want to replace those queries with insert queries with `returning`

nwalberts replied:
`userVotes` is both a poor name (since it should contain but one vote) and should exist on its own, not as a property of `newVote`. We don'y always need to assign our data to properties on one object to be passed back

nwalberts commented on server/src/routes/api/v1/podcastVotesRouter.js:
```diff
@@ -0,0 +1,105 @@
+      .where('podcastId', podcastId)
+      .sum('value as value')
+      .first()
+    return res.status(201).json({ votes: newVote })
```
you should not be assigning `totalVotes` as a property of `newVote`. totalVotes is relevant to a Podcast, note a vote (a vote shouldnt know the amount of total vote for a specific podcast). 
Instead, in your `.json`, return two poperties, `votes` and `totalVotes`

nwalberts commented on server/src/routes/api/v1/podcastVotesRouter.js:
```diff
@@ -0,0 +1,105 @@
+      .where('podcastId', podcastId)
+      .andWhere('userId', userId)
+      .first()
+      .returning('*')
```
again, same as above
- editVote is not needed, since userVotes and totalVotes can exist on their own
- userVotes is a poor name, since it will only contain one vote
- Not super keeon on this use of `where`. In fact, you should be using `updateAndFetchById` ideally (this would require holding the id of your vote in state on the react side). However, I will allow this the refactor process is too cumbersome (but I reserve the right to groan about it)

nwalberts commented on server/src/routes/api/v1/podcastVotesRouter.js:
```diff
@@ -0,0 +1,105 @@
+    editVote.totalVotes = await Vote.query()
+      .where('podcastId', podcastId)
+      .sum('value as value')
+      .first()
```
again, this `.first()` is odd, but if it actually returns the number value you need, Ill allow it

nwalberts commented on server/src/routes/api/v1/podcastVotesRouter.js:
```diff
@@ -0,0 +1,105 @@
+    deletedVote.totalVotes = await Vote.query()
+      .where('podcastId', podcastId)
+      .sum('value as value')
+      .first()
```
same as previous routes
- no need for deletedVote, userVotes and totalVotes can exist on their own 
- I really dont see the need for userVotes at all in fact, since the return of delete is the number of deleted rows
- total votes makes sense, but should not be a property of deletedVote

nwalberts replied:
i just checked, and I dont even think you use `userVotes` after your DELETE fetch request at all

nwalberts commented on client/src/components/podcasts/PodcastTile.js:
```diff
@@ -1,20 +1,167 @@
+    } else {
+      newVote({ value: -1 })
+    }
+  }
```
ill not require this, but these vote handlers have so much logic, that it would serve to make a seperate function or object that isnt a component to help manage this logic. but that would be an extensive refactor

nwalberts reviewed (changes requested):
Lots of changes requested, and many lines of code here that just don't line up. 

Please review my comments, take notes, test things out, refactor what you can, and then after refactoring those things that make sense, reach out to me if you would like to sync and discuss the rest. You may even respond to some of my comments in Github directly so that we can use them to discuss

tomlow committed:
Merge branch 'voting-system' of https://github.com/TurtIeSocks/launch31-group-project into voting-system

TurtIeSocks committed:
Simplifications

- Reduce # of fetch requests
- Cleanup of Objection code
- Remove uneeded relationships
